tests/TestHarness/launcher.py

```diff
@@ -66,9 +188,264 @@ def comma_separated(string):
+        i = 0
+        producer_number = 0
+        to_not_start_node = self.args.total_nodes - self.args.unstarted_nodes - 1
+        for node_name, node in self.network.nodes.items():
+            is_bios = node_name == 'bios'
+            node.keys.append(KeyStrings('EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV',
+                                         '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3'))
+            if is_bios:
+                node.producers.append('eosio')
+            else:
+                if i < non_bios:
+                    count = per_node
+                    if extra:
+                        count += 1
+                        extra -= 1
+                    while count:
+                        prodname = producer_name(producer_number)
+                        node.producers.append(prodname)
+                        producer_number += 1
+                        count -= 1
+                    for j in range(0, self.args.shared_producers):
+                        prodname = producer_name(j, True)
+                        node.producers.append(prodname)
+                node.dont_start = i >= to_not_start_node
+            if not is_bios:
+                i += 1
+
+
+    def generate(self):
+        {
+            'ring': self.make_line,
+            'line': lambda: self.make_line(False),
+            'star': self.make_star,
+            'mesh': self.make_mesh,
+        }.get(self.args.shape, self.make_custom)()
+
+        if not self.args.nogen:
+            for node_name, node in self.network.nodes.items():
+                node.config_dir_name.mkdir(parents=True)
+                self.write_config_file(node)
+                self.write_logging_config_file(node)
+                self.write_genesis_file(node)
+        
+        self.write_dot_file()
+
+    def write_config_file(self, node):
+        is_bios = node.name == 'bios'
+        peers = '\n'.join([f'p2p-peer-address = {self.network.nodes[p].p2p_endpoint}' for p in node.peers])
+        if len(node.producers) > 0:
+            producer_keys = 'private-key = ["EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV","5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3"]\n'
+            producer_names = '\n'.join([f'producer-name = {p}' for p in node.producers])
+            producer_plugin = 'plugin = eosio::producer_plugin\n'
+        else:
+            producer_keys = ''
+            producer_names = ''
+            producer_plugin = ''
+        config = f'''blocks-dir = {block_dir}
+http-server-address = {node.host_name}:{node.http_port}
+http-validate-host = false
+p2p-listen-endpoint = {node.listen_addr}:{node.p2p_port}
+p2p-server-address = {node.p2p_endpoint}
+{"enable-stale-production = true" if is_bios else ""}
+{"p2p-peer-address = %s" % self.network.nodes['bios'].p2p_endpoint if not is_bios else ""}
+{peers}
+{producer_keys}
+{producer_names}
+{producer_plugin}
+plugin = eosio::net_plugin
+plugin = eosio::chain_api_plugin
+'''
```

Comment:
Is there a way to factor this out or something to get it to look better format-wise?

Reply:
This _is_ the attempt to make it more readable.  The obvious is a series of `write()`s in the style of the C++ version.  The first attempt was using Python's built in `Template` class from the `string` module, but I consider `${block_dir}` to be less readable than f-string `{block_dir}`, plus Templates can't accept expressions between the braces.  The string in braces is solely a tag in the Template implementation, so the embedded conditionals aren't possible.  There are third party template libraries for Python (a dozen of them), many of which do support embedded logic, but I was trying to avoid another dependency.

Reply:
I agree that is this is very readable, I liked how you implemented that.  For some reason 0 indentations on the lines following the first just felt off to me.  I'm fine with this as is.